NSS: change default value of 'cache_first' to 'true'

alexey-tikhonov opened the pull request:
Having 'cache_first' as 'false' is a performance degradation without a reason
in the multi-(sub)domains enviroments (typical case when AD is involved).

:config: Default value of 'cache_first' option was changed to 'true'

alexey-tikhonov commented:
https://bugzilla.redhat.com/show_bug.cgi?id=1245367#c30

alexey-tikhonov commented:
Few integration tests are affected.

alexey-tikhonov commented:
It looks this change of default value uncovers real issues in the code.

Let's start with `test_ldap.py::test_lookup_by_email`.

With old default value of 'cache_first' the flow in NSS responder is:
```
[cache_req_send] (0x0400): CR #0: REQ_TRACE: New request [CID #1] 'User by name'
[cache_req_process_input] (0x0400): CR #0: Parsing input name [user1.email@LDAP]
[sss_parse_name_for_domains] (0x0200): name 'user1.email@LDAP' matched expression for domain 'LDAP', user is user1.email
[nss_get_object_send] (0x0400): Client [0x206f9b0][22]: sent cache request #0
[cache_req_set_name] (0x0400): CR #0: Setting name [user1.email]
[cache_req_select_domains] (0x0400): CR #0: Performing a single domain search
[sss_domain_get_state] (0x1000): Domain LDAP is Active
[cache_req_search_domains] (0x0400): CR #0: Search will check the cache and check the data provider
[cache_req_validate_domain_type] (0x2000): Request type POSIX-only for domain LDAP type POSIX is valid
[cache_req_set_domain] (0x0400): CR #0: Using domain [LDAP]
[cache_req_prepare_domain_data] (0x0400): CR #0: Preparing input data for domain [LDAP] rules
[cache_req_search_send] (0x0400): CR #0: Looking up user1.email@ldap
[cache_req_search_ncache] (0x0400): CR #0: Checking negative cache for [user1.email@ldap]
[sss_ncache_check_str] (0x2000): Checking negative cache for [NCE/USER/LDAP/user1.email@ldap]
[cache_req_search_ncache] (0x0400): CR #0: [user1.email@ldap] is not present in negative cache
[cache_req_search_cache] (0x0400): CR #0: Looking up [user1.email@ldap] in cache
[cache_req_search_cache] (0x0400): CR #0: Object [user1.email@ldap] was not found in cache
[cache_req_search_dp] (0x0400): CR #0: Looking up [user1.email@ldap] in data provider
[sss_dp_get_account_send] (0x0400): Creating request for [LDAP][0x1][BE_REQ_USER][name=user1.email@ldap:-]
[sbus_dispatch] (0x4000): Dispatching.
[sss_domain_get_state] (0x1000): Domain LDAP is Active
[cache_req_search_cache] (0x0400): CR #0: Looking up [user1.email@ldap] in cache
[cache_req_search_cache] (0x0400): CR #0: Object [user1.email@ldap] was not found in cache
[cache_req_search_ncache_add_to_domain] (0x0400): CR #0: Adding [user1.email@ldap] to negative cache
[sss_ncache_set_str] (0x0400): Adding [NCE/USER/LDAP/user1.email@ldap] to negative cache
[cache_req_set_plugin] (0x2000): CR #0: Setting "User by UPN" plugin
```
And then lookup by UPN succeeds.

But with `cache_first = true` it is:
```
[cache_req_send] (0x0400): CR #0: REQ_TRACE: New request [CID #1] 'User by name'
[cache_req_process_input] (0x0400): CR #0: Parsing input name [user1.email@LDAP]
[sss_parse_name_for_domains] (0x0200): name 'user1.email@LDAP' matched expression for domain 'LDAP', user is user1.email
[nss_get_object_send] (0x0400): Client [0x8e69b0][22]: sent cache request #0
[cache_req_set_name] (0x0400): CR #0: Setting name [user1.email]
[cache_req_select_domains] (0x0400): CR #0: Performing a single domain search
[sss_domain_get_state] (0x1000): Domain LDAP is Active
[cache_req_search_domains] (0x0400): CR #0: Search will check the cache and bypass the data provider
[cache_req_validate_domain_type] (0x2000): Request type POSIX-only for domain LDAP type POSIX is valid
[cache_req_set_domain] (0x0400): CR #0: Using domain [LDAP]
[cache_req_prepare_domain_data] (0x0400): CR #0: Preparing input data for domain [LDAP] rules
[cache_req_search_send] (0x0400): CR #0: Looking up user1.email@ldap
[cache_req_search_ncache] (0x0400): CR #0: Checking negative cache for [user1.email@ldap]
[sss_ncache_check_str] (0x2000): Checking negative cache for [NCE/USER/LDAP/user1.email@ldap]
[cache_req_search_ncache] (0x0400): CR #0: [user1.email@ldap] is not present in negative cache
[cache_req_search_cache] (0x0400): CR #0: Looking up [user1.email@ldap] in cache
[cache_req_search_cache] (0x0400): CR #0: Object [user1.email@ldap] was not found in cache
[cache_req_select_domains] (0x0400): CR #0: Performing a single domain search
[sss_domain_get_state] (0x1000): Domain LDAP is Active
[cache_req_search_domains] (0x0400): CR #0: Search will bypass the cache and check the data provider
[cache_req_validate_domain_type] (0x2000): Request type POSIX-only for domain LDAP type POSIX is valid
[cache_req_set_domain] (0x0400): CR #0: Using domain [LDAP]
[cache_req_prepare_domain_data] (0x0400): CR #0: Preparing input data for domain [LDAP] rules
[cache_req_search_send] (0x0400): CR #0: Looking up user1.email@ldap
[cache_req_search_ncache] (0x0400): CR #0: Checking negative cache for [user1.email@ldap]
[sss_ncache_check_str] (0x2000): Checking negative cache for [NCE/USER/LDAP/user1.email@ldap]
[cache_req_search_ncache] (0x0400): CR #0: [user1.email@ldap] is not present in negative cache
[cache_req_search_dp] (0x0400): CR #0: Looking up [user1.email@ldap] in data provider
[sss_dp_get_account_send] (0x0400): Creating request for [LDAP][0x1][BE_REQ_USER][name=user1.email@ldap:-]
[sbus_dispatch] (0x4000): Dispatching.
[sss_domain_get_state] (0x1000): Domain LDAP is Active
[cache_req_search_cache] (0x0400): CR #0: Looking up [user1.email@ldap] in cache
[cache_req_search_cache] (0x0400): CR #0: Object [user1.email@ldap] was not found in cache
[cache_req_search_ncache_add_to_domain] (0x0400): CR #0: Adding [user1.email@ldap] to negative cache
[sss_ncache_set_str] (0x0400): Adding [NCE/USER/LDAP/user1.email@ldap] to negative cache
[cache_req_process_result] (0x0400): CR #0: Finished: Not found
```
  --  it doesn't continue with lookup by UPN.

@pbrezina, why so?

pbrezina commented:
Looks like a bug. Try to catch it in gdb.

alexey-tikhonov commented:
> Looks like a bug. Try to catch it in gdb.

I don't have a capacity to debug this now, but I was able to reproduce it without unit test, so opened #6059

alexey-tikhonov commented:
> # Cppcheck Result
> 
> Clever!

@ikerexxe, @pbrezina,
do we really need a comments in a PR?
Wouldn't it be enough to have result in "checks list"?

ikerexxe commented:
> @ikerexxe, @pbrezina, do we really need bots comments in a PR? Wouldn't it be enough to have result in "checks list"?

I kind of agree that this comments can spam the comment workflow. I've opened a [PR](https://github.com/linuxdeepin/action-cppcheck/pull/9) in the action repository to make them optional. Once it's merged I'll repoint the commit ID from our workflow and disable the comments to avoid this type of messages.

alexey-tikhonov commented:
> > @ikerexxe, @pbrezina, do we really need bots comments in a PR? Wouldn't it be enough to have result in "checks list"?
> 
> I kind of agree that this comments can spam the comment workflow. I've opened a [PR](https://github.com/linuxdeepin/action-cppcheck/pull/9) in the action repository to make them optional. Once it's merged I'll repoint the commit ID from our workflow and disable the comments to avoid this type of messages.

Thanks!

alexey-tikhonov commented:
Fixing #6059 resolved an issues with `test_ldap.py::test_lookup_by_email`
But a lot of `test_files_provider.py` still fail.

pbrezina commented:
Hi Alexey, is this still WIP or is it waiting for a review?

alexey-tikhonov commented:
> Hi Alexey, is this still WIP or is it waiting for a review?

It's waiting for somebody (I guess me) to figure out why this setting makes 'files provider' tests to fail...
Should be done for 2.8.0 release.

alexey-tikhonov commented:
Backend:
```
(17:49:57): [process_dir_event] (0x0400): received notification for watched file [passwd] under ...
(17:49:57): [dp_sbus_domain_inconsistent] (0x0400): Ordering responders to disable domain files
(17:49:57): [sf_enum_files_steps] (0x4000): Step WAIT_TO_START_USERS.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:57): [sbus_dispatch] (0x4000): Dispatching.
(17:49:58): [sbus_dispatch] (0x4000): Dispatching.
(17:49:58): [sbus_dispatch] (0x4000): Dispatching.
(17:49:58): [sbus_method_handler] (0x2000): Received D-Bus method sssd.dataprovider.getAccountInfo on /sssd
(17:49:58): [sbus_senders_lookup] (0x2000): Looking for identity of sender [sssd.nss]
(17:49:58): [dp_get_account_info_send] (0x0200): Got request for [0x1][BE_REQ_USER][*]
(17:49:58): [dp_attach_req] (0x0400): [RID#1] DP Request [Account #1]: REQ_TRACE: New request. [sssd.nss CID #1] Flags [0x0001].
(17:49:58): [dp_attach_req] (0x0400): [RID#1] Number of active DP request: 1
(17:49:58): [sss_domain_get_state] (0x1000): [RID#1] Domain files is Active
(17:49:58): [sf_enum_files_steps] (0x4000): Step WAIT_TO_START_USERS.
(17:49:58): [sf_enum_files_steps] (0x4000): Step DELETE_USERS.
(17:49:58): [sysdb_delete_recursive_with_filter] (0x4000): Found [1] items to delete.
(17:49:58): [sysdb_delete_recursive_with_filter] (0x4000): Trying to delete [name=canary@files,cn=users,cn=files,cn=sysdb].
(17:49:58): [sf_enum_files_steps] (0x4000): Step READ_USERS.
(17:49:58): [enum_files_users] (0x1000): User found (root, x, 0, 0, root, /home/talex/src/sssd/x86_64/intg/bld/src/tests/intg/root, /bin/bash)
(17:49:58): [enum_files_users] (0x1000): User found (canary, x, 100001, 200001, Used to check if passwd is resolvable, /home/canary, /bin/bash)
(17:49:58): [enum_files_users] (0x1000): User found (user1, x, 10001, 20001, , /home/user1, /bin/bash)
(17:49:58): [sf_enum_files_steps] (0x4000): Step SAVE_USERS.
(17:49:58): [save_file_user] (0x0400): Skipping root
(17:49:58): [sysdb_search_by_name] (0x0400): No such entry
(17:49:58): [sysdb_store_user] (0x1000): User canary@files does not exist.
(17:49:58): [sysdb_search_user_by_uid] (0x0400): No such entry
(17:49:58): [sysdb_ldb_msg_difference] (0x2000): Added attr [isPosix] to entry [name=canary@files,cn=users,cn=files,cn=sysdb]
(17:49:58): [sysdb_set_entry_attr] (0x0200): Entry [name=canary@files,cn=users,cn=files,cn=sysdb] has set [cache, ts_cache] attrs.
(17:49:58): [sysdb_store_user] (0x0400): User "canary@files" has been stored
(17:49:58): [sysdb_search_by_name] (0x0400): No such entry
(17:49:58): [sysdb_store_user] (0x1000): User user1@files does not exist.
(17:49:58): [sysdb_search_user_by_uid] (0x0400): No such entry
(17:49:58): [sysdb_ldb_msg_difference] (0x2000): Added attr [isPosix] to entry [name=user1@files,cn=users,cn=files,cn=sysdb]
(17:49:58): [sysdb_set_entry_attr] (0x0200): Entry [name=user1@files,cn=users,cn=files,cn=sysdb] has set [cache, ts_cache] attrs.
(17:49:58): [sysdb_store_user] (0x0400): User "user1@files" has been stored
(17:49:58): [refresh_override_attrs] (0x0400): No overrides, nothing to do.
(17:49:58): [sf_enum_files_steps] (0x4000): Step READ_USERS.
(17:49:58): [sf_enum_files_steps] (0x4000): Step WAIT_TO_START_GROUPS.
(17:49:58): [sf_enum_files_steps] (0x4000): Step DELETE_GROUPS.
(17:49:58): [sf_enum_files_steps] (0x4000): Step READ_GROUPS.
(17:49:58): [enum_files_groups] (0x1000): Group found (root, 0)
(17:49:58): [sf_enum_files_steps] (0x4000): Step SAVE_GROUPS.
(17:49:58): [sysdb_enumpwent_filter] (0x1000): Searching timestamp cache with [(objectCategory=user)]
(17:49:58): [sysdb_cache_search_users] (0x2000): Search users with filter: (&(objectCategory=user)(objectCategory=user))
(17:49:58): [sysdb_enumpwent_filter] (0x1000): Searching timestamp entries with [(|(dn=name=canary@files,cn=users,cn=files,cn=sysdb)(dn=name=user1@files,cn=users,cn=files,cn=sysdb))]
(17:49:58): [sysdb_enumpwent_filter] (0x1000): Searching cache with [(objectCategory=user)]
(17:49:58): [save_file_group] (0x0400): Skipping root
(17:49:58): [refresh_override_attrs] (0x0400): No overrides, nothing to do.
(17:49:58): [sf_enum_files_steps] (0x4000): Step READ_GROUPS.
(17:49:58): [sf_enum_files_steps] (0x4000): Step UPDATE_FINISH.
(17:49:58): [sysdb_cache_search_users] (0x2000): Search users with filter: (&(objectCategory=user)(name=*))
(17:49:58): [sss_domain_get_state] (0x1000): Domain files is Active
(17:49:58): [sss_domain_get_state] (0x1000): Domain files is Active
(17:49:58): [dp_sbus_domain_active] (0x0400): Ordering responders to enable domain files
(17:49:58): [sbus_requests_add] (0x4000): Chaining request: -:0:sssd.Responder.Domain.SetActive:/sssd:files
(17:49:58): [sbus_requests_add] (0x4000): Chaining request: -:0:sssd.Responder.Domain.SetActive:/sssd:files
(17:49:58): [sbus_requests_add] (0x4000): Chaining request: -:0:sssd.Responder.Domain.SetActive:/sssd:files
(17:49:58): [sbus_requests_add] (0x4000): Chaining request: -:0:sssd.Responder.Domain.SetActive:/sssd:files
(17:49:58): [sbus_requests_add] (0x4000): Chaining request: -:0:sssd.Responder.Domain.SetActive:/sssd:files
(17:49:58): [sbus_requests_add] (0x4000): Chaining request: -:0:sssd.Responder.Domain.SetActive:/sssd:files
(17:49:58): [dp_req_done] (0x0400): DP Request [Account #1]: Request handler finished [0]: Success
(17:49:58): [dp_req_done] (0x20000): DP Request [Account #1]: Handling request took [14.702] milliseconds.
(17:49:58): [_dp_req_recv] (0x0400): DP Request [Account #1]: Receiving request data.
(17:49:58): [dp_req_destructor] (0x0400): DP Request [Account #1]: Request removed.
(17:49:58): [dp_req_destructor] (0x0400): Number of active DP request: 0
(17:49:58): [dp_req_reply_std] (0x1000): DP Request [Account #1]: Returning [Success]: 0,0,(null)
(17:49:58): [sbus_issue_request_done] (0x0400): sssd.dataprovider.getAccountInfo: Success
```

alexey-tikhonov commented:
(1) why 'files' provider is stuck on `Step WAIT_TO_START_USERS` (until it receives `[BE_REQ_USER][*]`)?

(2) is there a logical loop:
```
(17:49:58): [nss] [sbus_method_handler] (0x2000): Received D-Bus method sssd.Responder.Domain.SetActive on /sssd
(17:49:58): [nss] [sbus_senders_lookup] (0x2000): Looking for identity of sender [sssd.domain_files]
(17:49:58): [nss] [sbus_dispatch] (0x4000): Dispatching.
(17:49:58): [nss] [sss_domain_get_state] (0x1000): [CID#1] Domain files is Inconsistent
(17:49:58): [nss] [cache_req_search_done] (0x0400): [CID#1] CR #0: Domain inconsistent, we will not return cached data
```
  --  looks like `sssd_nss`, that requested update, can't process `SetActive` in this state?

@sumit-bose , @pbrezina , please take a look.

sumit-bose commented on src/responder/common/responder_common.c:
```diff
@@ -1341,7 +1341,15 @@ int sss_process_init(TALLOC_CTX *mem_ctx,
     ret = confdb_get_bool(rctx->cdb, rctx->confdb_service_path,
                           CONFDB_RESPONDER_CACHE_FIRST,
-                          false, &rctx->cache_first);
+#ifdef BUILD_FILES_PROVIDER
```
Hi,

I think it would be more consistent with reading other confdb parametersto define a `CONFDB_RESPONDER_CACHE_FIRST_DEFAULT` in `src/confdb/confdb.h` and move the `#ifdef` there.

bye,
Sumit

alexey-tikhonov replied:
Thanks, done.

pbrezina reviewed (approved):
Thank you. Ack.

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/6052

* `master`
    * ae691f0b415ad0bcf09a39d0272331fd74d65299 - NSS: change default value of 'cache_first' to 'true'
